lib/apiservers/engine/backends/cache/image_cache.go

```diff
@@ -115,3 +119,51 @@ func (c *ImageCache) GetImages() []*metadata.ImageConfig {
 
 	return result
 }
+
+func (c *ImageCache) GetImageByDigest(digest digest.Digest) *metadata.ImageConfig {
+	c.m.RLock()
+	defer c.m.RUnlock()
+
+	var imageID string
+
+	for _, image := range c.cache {
+		if strings.Contains(image.ImageID, "sha") {
+			imageID = image.ImageID
+		} else {
+			imageID = "sha256:" + image.ImageID
+		}
+
+		if imageID == string(digest) {
+			newImage := new(metadata.ImageConfig)
```

Comment:
If this is attempting to isolate the cache from changes then it needs to do a deep copy. At the moment changing elements in the history array in the returned structure will alter the entry in the cache:
https://play.golang.org/p/By21rCOeg1